**Weather Data set**

A Time-Series data set 
with per hour information
about the weather conditions at a particular location.
Records Temp, Dew point temp, Relative Humidity, Wind speed, Visibility, Pressure and Conditions

CSV file

**Notes**

head() - It shows the first N rows in the data (by default, N=5).

shape - It shows the total no. of rows and no. of columns of the dataframe.

index - This attribute provides the index of the dataframe.

columns - It shows the name of each column.

dtypes - It shows the data-type of each column.

unique() - In a column, it shows all the unique values. It can be applied on a single column only, not on the whole dataframe.

nunique() - It shows the total no. of unique values in each column. It can be applied on a single column as well as on the whole dataframe.

count - It shows the total no. of non-null values in each column. It can be applied on a single column as well as on the whole dataframe.

value_counts - In a column, it shows all the unique values with their count. It can be applied on a single column only.

info() - Provides basic information about the dataframe.

In [3]:
import pandas as pd

In [10]:
data = pd.read_csv(r"C:\Users\hp\OneDrive\Desktop\datasets\weatherData.csv")

r is for removing unicode error

In [15]:
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

8784 rows

8 columns

Date/Time, Temp_C,	Dew Point, Temp_C,	Rel Hum_%,	Wind Speed_km/h, Visibility_km,	Press_kPa.	Weather

head() - It shows the first N rows in the data (by default, N=5).


In [23]:
data.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

tail() - last 5 data

In [26]:
data.tail()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather  
8779            9.7     100.13    Snow  
8780            9.7     100.03    Snow  
8781            4.8      99.95    Snow  
8782            9.7      99.91    Snow  
8783           11.3      99.89    Snow

.shape - It shows the total no. of rows and no. of columns of the dataframe.

In [38]:
data.shape

(8784, 8)

index of the data frame
starting from 0 and last position is 8784 and increment by 1

In [43]:
data.index

RangeIndex(start=0, stop=8784, step=1)

In [45]:
data.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

In [47]:
data.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

unique data values present in weather column

In [53]:
data['Weather'].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

In [56]:
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

total no. of non null val in each column. can apply on single column or whole dataset

In [70]:
data.count

<bound method DataFrame.count of              Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      

In [82]:
data['Weather'].count

<bound method Series.count of 0                        Fog
1                        Fog
2       Freezing Drizzle,Fog
3       Freezing Drizzle,Fog
4                        Fog
                ...         
8779                    Snow
8780                    Snow
8781                    Snow
8782                    Snow
8783                    Snow
Name: Weather, Length: 8784, dtype: object>

.value_counts
unique values with their count
applied on single columns only

In [85]:
data['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [87]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


Unique wind speed values in the data

In [108]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [110]:
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

In [114]:
data['Wind Speed_km/h'].nunique()

34

In [116]:
data['Wind Speed_km/h'].count()

8784

In [118]:
data['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2],
      dtype=int64)

number of times when the Weather is exactly clear

using Value count

In [121]:
data.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

In [123]:
data['Weather'].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

In [152]:
data['Weather'].value_counts()

Weather
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                

In [169]:
clear = data['Weather'] == 'Clear'

In [177]:
clear_count = clear.value_counts()

In [183]:
clear_count

Weather
False    7458
True     1326
Name: count, dtype: int64

In [185]:
clear_count[True]

1326

*Using Filtering*

In [203]:
data[data.Weather == 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

*using groupby*

In [206]:
data.groupby('Weather').get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

Number of times when the wind speed was exactly 4km/h

In [209]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [229]:
windSpeed = data['Wind Speed_km/h'] == 4

In [239]:
speed_4 = windSpeed.value_counts()

In [241]:
speed_4

Wind Speed_km/h
False    8310
True      474
Name: count, dtype: int64

In [253]:
speed_4[True]

474

In [255]:
data[data['Wind Speed_km/h'] == 4]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

In [259]:
data.groupby('Wind Speed_km/h').get_group(4)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

All the null values in the data

In [264]:
data.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [266]:
data.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

Rename the column name Weather in to Weather Condition

In [271]:
data.rename(columns = {'Weather' : 'Weather Condition'})

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [275]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog

In [279]:
data.rename(columns = {'Weather' : 'Weather Conditions'}, inplace = True)

In [281]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather Conditions  
0            8.0     101.24                Fog

Mean Visibility

In [286]:
data['Visibility_km'].mean()

27.664446721311478

Standerd Deviation of Pressure

In [291]:
data.Press_kPa.std()

0.8440047459486474

In [296]:
#if there is no space okay to use after . no need of []

Varioance of Relative Humidity

In [299]:
data['Rel Hum_%'].var()

286.2485501984998

All instances when Snow was recorded

In [302]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather Conditions  
0            8.0     101.24                Fog

In [310]:
 data['Weather Conditions'].value_counts()

Weather Conditions
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle     

In [314]:
snow = data['Weather Conditions'] == 'Snow'

In [334]:
snow_count = snow.value_counts()

In [336]:
snow_count[True]

390

In [338]:
data.groupby('Weather Conditions').get_group('Snow')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Conditions  
55             25.0     100.95               Snow  
84             24.1     101.25               Snow  
86             19.3     100.97               Snow  
87              9.7     100.89               Snow  
88             19.3     100.79               Snow  
...             ...        ...                ...  
8779            9.7     100.13               Snow  
8780            9.7     100.03               Snow  
8781            4.8      99.95               Snow  
8782            9.7      99.91               Snow  
8783           11.3      99.89               Snow  

[390 rows x 8 columns]

all instances when wind speed is above 24 and visibility is 25

In [408]:
windSpd = data[data['Wind Speed_km/h'] > 24]

In [410]:
vis = data[data['Visibility_km'] == 25]

In [412]:
data[(data['Wind Speed_km/h'] > 24) & (data['Visibility_km'] == 25)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa Weather Conditions  
23             25.0      99.31             Cloudy  
24             25.0      99.26       Rain Showers  
25             25.0      99.26             Cloudy  
26             25.0      99.26      Mostly Cloudy  
27             25.0      99.30      Mostly Cloudy  
...             ...        ...                ...  
8705           25.0     101.34       Mainly Clear  
8753           25.0     101.26       Mainly Clear  
8755           25.0     101.47       Mainly Clear  
8759           25.0     101.52      Mostly Cloudy  
8760           25.0     101.51             Cloudy  

[308 rows x 8 columns]

mean value of each column against each 'weather condition'

In [415]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather Conditions  
0            8.0     101.24                Fog

In [421]:
data['Weather Conditions']

0                        Fog
1                        Fog
2       Freezing Drizzle,Fog
3       Freezing Drizzle,Fog
4                        Fog
                ...         
8779                    Snow
8780                    Snow
8781                    Snow
8782                    Snow
8783                    Snow
Name: Weather Conditions, Length: 8784, dtype: object

In [425]:
data.groupby('Weather Conditions').mean(numeric_only=True)

Temp_C  Dew Point Temp_C  \
Weather Conditions                                                     
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather Conditions                                                    
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        

min and max val of each cell against weather condition

In [428]:
data.groupby('Weather Conditions').min(numeric_only=True)

Temp_C  Dew Point Temp_C  Rel Hum_%  \
Weather Conditions                                                             
Clear                                     -23.3             -28.5         20   
Cloudy                                    -21.4             -26.8         18   
Drizzle                                     1.1              -0.2         74   
Drizzle,Fog                                 0.0              -1.6         85   
Drizzle,Ice Pellets,Fog                     0.4              -0.7         92   
Drizzle,Snow                                0.9               0.1         92   
Drizzle,Snow,Fog                            0.3              -0.1         92   
Fog                                       -16.0             -17.2         80   
Freezing Drizzle                           -9.0             -12.2         78   
Freezing Drizzle,Fog                       -6.4              -9.0         82   
Freezing Drizzle,Haze                      -5.8              -8.3         81   
Freezing Drizzle,Snow                      -8.3             -10.4         79   
Freezing Fog                              -19.0             -22.9         71   
Freezing Rain                              -6.5              -9.0         81   
Freezing Rain,Fog                          -6.1              -8.7         82   
Freezing Rain,Haze                         -4.9              -7.5         82   
Freezing Rain,Ice Pellets,Fog              -2.6              -3.7         92   
Freezing Rain,Snow Grains                  -5.0              -7.3         84   
Haze                                      -11.5             -16.0         68   
Mainly Clear                              -22.8             -28.0         20   
Moderate Rain,Fog                           1.7               0.8         94   
Moderate Snow                              -6.3              -7.6         83   
Moderate Snow,Blowing Snow                 -5.5              -6.6         92   
Mostly Cloudy                             -23.2             -28.5         18   
Rain                                        0.3              -5.7         40   
Rain Showers                                1.6              -7.2         37   
Rain Showers,Fog                           12.8              12.1         96   
Rain Showers,Snow Showers                   2.1              -1.8         75   
Rain,Fog                                    0.0              -1.2         83   
Rain,Haze                                   4.0               1.0         81   
Rain,Ice Pellets                            0.6              -0.6         92   
Rain,Snow                                   0.6              -1.7         81   
Rain,Snow Grains                            1.9              -2.1         75   
Rain,Snow,Fog                               0.8               0.3         96   
Rain,Snow,Ice Pellets                       0.9              -0.7         88   
Snow                                      -16.7             -24.6         41   
Snow Pellets                                0.7              -6.4         59   
Snow Showers                              -13.3             -19.3         52   
Snow Showers,Fog                          -11.3             -12.7         89   
Snow,Blowing Snow                         -12.0             -16.2         70   
Snow,Fog                                  -10.1             -12.0         77   
Snow,Haze                                  -4.3              -7.2         80   
Snow,Ice Pellets                           -4.3              -5.9         76   
Thunderstorms                              21.6              19.4         67   
Thunderstorms,Heavy Rain Showers           10.9               9.0         88   
Thunderstorms,Moderate Rain Showers,Fog    19.6              18.5         93   
Thunderstorms,Rain                         19.4              18.2         83   
Thunderstorms,Rain Showers                 11.0               7.0         68   
Thunderstorms,Rain Showers,Fog           

In [432]:
data.groupby('Weather Conditions').max(numeric_only=True)

Temp_C  Dew Point Temp_C  Rel Hum_%  \
Weather Conditions                                                             
Clear                                      32.8              20.4         99   
Cloudy                                     30.5              22.6         99   
Drizzle                                    18.8              17.7         96   
Drizzle,Fog                                19.9              19.1        100   
Drizzle,Ice Pellets,Fog                     0.4              -0.7         92   
Drizzle,Snow                                1.2               0.2         95   
Drizzle,Snow,Fog                            1.1               0.6         98   
Fog                                        20.8              19.6        100   
Freezing Drizzle                           -2.3              -3.3         93   
Freezing Drizzle,Fog                       -0.3              -2.3         94   
Freezing Drizzle,Haze                      -5.0              -7.7         83   
Freezing Drizzle,Snow                      -3.3              -4.6         94   
Freezing Fog                               -0.1              -0.3         99   
Freezing Rain                               0.3              -1.7         92   
Freezing Rain,Fog                           0.1              -0.9         93   
Freezing Rain,Haze                         -4.9              -7.4         83   
Freezing Rain,Ice Pellets,Fog              -2.6              -3.7         92   
Freezing Rain,Snow Grains                  -5.0              -7.3         84   
Haze                                       14.1              11.1         86   
Mainly Clear                               33.0              21.2         99   
Moderate Rain,Fog                           1.7               0.8         94   
Moderate Snow                              -4.9              -6.7         93   
Moderate Snow,Blowing Snow                 -5.4              -6.4         93   
Mostly Cloudy                              32.4              24.4        100   
Rain                                       22.8              20.4         99   
Rain Showers                               26.4              23.0         97   
Rain Showers,Fog                           12.8              12.1         96   
Rain Showers,Snow Showers                   2.2              -1.2         78   
Rain,Fog                                   21.7              19.5        100   
Rain,Haze                                   5.5               2.9         86   
Rain,Ice Pellets                            0.6              -0.6         92   
Rain,Snow                                   1.7               0.5         94   
Rain,Snow Grains                            1.9              -2.1         75   
Rain,Snow,Fog                               0.8               0.3         96   
Rain,Snow,Ice Pellets                       1.3               0.1         94   
Snow                                        3.7               0.3         96   
Snow Pellets                                0.7              -6.4         59   
Snow Showers                                2.9              -0.7         94   
Snow Showers,Fog                          -10.0             -11.1         92   
Snow,Blowing Snow                          -1.4              -2.9         91   
Snow,Fog                                    1.1               0.8         99   
Snow,Haze                                  -3.6              -6.4         81   
Snow,Ice Pellets                            0.8              -1.7         92   
Thunderstorms                              26.7              20.1         87   
Thunderstorms,Heavy Rain Showers           10.9               9.0         88   
Thunderstorms,Moderate Rain Showers,Fog    19.6              18.5         93   
Thunderstorms,Rain                         21.3              19.1         93   
Thunderstorms,Rain Showers                 25.5              23.1         98   
Thunderstorms,Rain Showers,Fog           

Show all the records of weather condition is Fog

In [445]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather Conditions  
0            8.0     101.24                Fog

In [451]:
data[data['Weather Conditions'] == 'Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Conditions  
0               8.0     101.24                Fog  
1               8.0     101.24                Fog  
4               4.8     101.23                Fog  
5               6.4     101.27                Fog  
6               6.4     101.29                Fog  
...             ...        ...                ...  
8716            9.7     101.25                Fog  
8717            6.4     101.25                Fog  
8718            9.7     101.25                Fog  
8719            8.0     101.22                Fog  
8722            6.4     101.15                Fog  

[150 rows x 8 columns]

In [453]:
data.groupby('Weather Conditions').get_group('Fog')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Conditions  
0               8.0     101.24                Fog  
1               8.0     101.24                Fog  
4               4.8     101.23                Fog  
5               6.4     101.27                Fog  
6               6.4     101.29                Fog  
...             ...        ...                ...  
8716            9.7     101.25                Fog  
8717            6.4     101.25                Fog  
8718            9.7     101.25                Fog  
8719            8.0     101.22                Fog  
8722            6.4     101.15                Fog  

[150 rows x 8 columns]

all records when weather is clear and visibility is above 40

In [458]:
data[(data['Weather Conditions']=='Clear') & (data['Visibility_km'] > 40)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
351    1/15/2012 15:00   -15.4             -22.8         53               24   
352    1/15/2012 16:00   -15.1             -22.8         52               24   
425    1/18/2012 17:00   -11.3             -18.8         54               26   
440     1/19/2012 8:00   -13.7             -18.4         68               19   
441     1/19/2012 9:00   -12.7             -17.2         69               17   
...                ...     ...               ...        ...              ...   
8384   12/15/2012 8:00   -10.7             -15.6         67               13   
8385   12/15/2012 9:00   -10.4             -15.9         64               19   
8389  12/15/2012 13:00    -8.4             -14.7         60               19   
8631  12/25/2012 15:00    -7.1             -13.7         59               17   
8632  12/25/2012 16:00    -7.5             -13.9         60               11   

      Visibility_km  Press_kPa Weather Conditions  
351            48.3     102.71              Clear  
352            48.3     102.79              Clear  
425            48.3     101.54              Clear  
440            48.3     101.84              Clear  
441            48.3     101.73              Clear  
...             ...        ...                ...  
8384           48.3     102.69              Clear  
8385           48.3     102.74              Clear  
8389           48.3     102.64              Clear  
8631           48.3     101.98              Clear  
8632           48.3     102.03              Clear  

[313 rows x 8 columns]

weather clear and relative humidity > 50

In [472]:
data[(data['Weather Conditions'] == 'Clear') & ((data['Rel Hum_%'] > 50) | (data['Visibility_km'] < 40))]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather Conditions  
67             25.0     101.74              Clear  
114            25.0     100.71              Clear  
115            25.0     100.80              Clear  
116            25.0     100.83              Clear  
117            25.0     100.83              Clear  
...             ...        ...                ...  
8646           25.0     102.47              Clear  
8698           24.1     101.27              Clear  
8713           25.0     101.31              Clear  
8714           25.0     101.33              Clear  
8756           25.0     101.52              Clear  

[1176 rows x 8 columns]

In [476]:
data.rename(columns = {'Rel Hum_%' : 'rel_humdity'})

Date/Time  Temp_C  Dew Point Temp_C  rel_humdity  \
0        1/1/2012 0:00    -1.8              -3.9           86   
1        1/1/2012 1:00    -1.8              -3.7           87   
2        1/1/2012 2:00    -1.8              -3.4           89   
3        1/1/2012 3:00    -1.5              -3.2           88   
4        1/1/2012 4:00    -1.5              -3.3           88   
...                ...     ...               ...          ...   
8779  12/31/2012 19:00     0.1              -2.7           81   
8780  12/31/2012 20:00     0.2              -2.4           83   
8781  12/31/2012 21:00    -0.5              -1.5           93   
8782  12/31/2012 22:00    -0.2              -1.8           89   
8783  12/31/2012 23:00     0.0              -2.1           86   

      Wind Speed_km/h  Visibility_km  Press_kPa    Weather Conditions  
0                   4            8.0     101.24                   Fog  
1                   4            8.0     101.24                   Fog  
2                   7            4.0     101.26  Freezing Drizzle,Fog  
3                   6            4.0     101.27  Freezing Drizzle,Fog  
4                   7            4.8     101.23                   Fog  
...               ...            ...        ...                   ...  
8779               30            9.7     100.13                  Snow  
8780               24            9.7     100.03                  Snow  
8781               28            4.8      99.95                  Snow  
8782               28            9.7      99.91                  Snow  
8783               30           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [478]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   

   Visibility_km  Press_kPa Weather Conditions  
0            8.0     101.24                Fog

In [480]:
data.rename(columns = {'Rel Hum_%' : 'Relative_humidity'}, inplace=True)

In [482]:
data.head(1)

Date/Time  Temp_C  Dew Point Temp_C  Relative_humidity  \
0  1/1/2012 0:00    -1.8              -3.9                 86   

   Wind Speed_km/h  Visibility_km  Press_kPa Weather Conditions  
0                4            8.0     101.24                Fog